Merge custom s3 bucket templates with the same bucket name. fixes #2717

asprouse opened the pull request:
<!--
1. Please check out and follow our Contributing Guidelines: https://github.com/serverless/serverless/blob/master/CONTRIBUTING.md
2. Do not remove any section of the template. If something is not applicable leave it empty but leave it in the PR
3. Please follow the template, otherwise we'll have to ask you to update it and it will take longer until your PR is merged
-->

## What did you implement:

Closes #2717

<!--
Briefly describe the feature if no issue exists for this PR
-->

## How did you implement it:

I updated `mergeCustomProviderResources.js` to merge s3 bucket templates based on bucket name. If a custom bucket template exists in the custom resources section of `serverless.yml` and the same bucket is referenced in a function event the two are merged.

## How can we verify it:

Unit tests

1. Create a function with an s3 event in `serverless.yml`
1. Add an s3 bucket in the `resources` section of `serverless.yml`
1. Add custom configuration to the bucket in `resources`. For example add CORS settings.
1. Run `sls deploy -v`
1. Verify the s3 bucket template was properly merged in`.serverless/cloudformation-template-update-stack.json` with the logical id of the custom resource.
1. Verify that the s3 bucket is created with the merged config.

## Todos:

- [x] Write tests
- [x] Write documentation
- [x] Fix linting errors
- [x] Make sure code coverage hasn't dropped
- [x] Provide verification config/commands/resources
- [x] Enable ["Allow edits from maintainers"](https://help.github.com/articles/allowing-changes-to-a-pull-request-branch-created-from-a-fork/) for this PR
- [x] Change ready for review message below


***Is this ready for review?:*** YES

asprouse commented:
Unsure of what sort of documentation to add.

pmuens commented:
Hey @asprouse thanks for opening this PR! 👍 

I added a comment in the corresponding issue (#2717) as your problem should be able to be resolved through the `resources.Resouces` section.

Let us know if the proposed solution works for you or if it still persists.

asprouse commented:
@pmuens It's been almost two months and #2717 is still open, and this PR is a solution. What needs to be done to get this merged?

pmuens commented:
Hey @asprouse thanks for pinging and sorry for the slow response on our end.

We'll look through all the open PRs starting next week so that we can assign specific PRs to milestone (and therefore releases).

We'll get back to this PR once we've gone through the process.

Thanks again for your work here 👍

nachoab commented:
FWIW I've been using this PR for a while otherwise I cannot deploy my s3 event + s3 resource definition

patoncrispy commented:
Any updates on this beauty? Would be great to have this feature available! 😃

asprouse commented:
Just rebased onto master and fixed the merge issue and added some basic docs. @pmuens @eahefnawy It seems like there is a lot of people depending on this fork. This PR has been languishing for almost 4 months. Please let me know what I can do to get this merged.

eahefnawy commented:
@asprouse really sorry, but I'm reluctant to merge this PR, mainly due to technical debt for a minor addition, which could be solved by overwriting our core S3 resource with custom resources templates. I completely understand your use case, but we want to keep the framework focused on Serverless, and not be a syntactic sugar on top of S3/CF, because otherwise we'd have to support a lot more S3/CF features than what this PR proposes. This would be a big distraction for us.

patoncrispy commented:
What does 'focused on Serverless' mean? Without this PR, it seems that support for S3 events is conflicted and somewhat lacking. While my Lambda function event will create the bucket required, this is basically useless as I am unable to configure the permissions required on the bucket using CF. CF is supported by Serverless, but in this case, the support is incomplete as it doesn't merge properly with my function event definitions (or vice-versa). 

I am not trying to be facetious or obtuse, but just trying to better understand the justification for closing this PR.

nachoab commented:
I understand your idea of not getting too deep into provider configurations while keeping it versatile via resources. But in this case creating the bucket automatically makes it impossible to configure it via SLS. If you are not going to make it work completely, then leave it so it doesn't fail straight away

eahefnawy commented:
@patoncrispy @nachoab you can always [overwrite our core resoruces with any config you want using custom resources](https://serverless.com/framework/docs/providers/aws/guide/resources/#override-aws-cloudformation-resource). We're just worried about the technical debt of supporting very specific/custom use case that you can already achieve by overwriting. I hope you guys understand technical debt slows us down a lot 😞

nachoab commented:
Configuring a lambda s3 event and updating the bucket using resources looks like a common use case, 
What about not creating the s3 bucket automatically when describing a s3 event?

HyperBrain commented:
@nachoab You mean that one declares his bucket like any other resource in the resources section and then just uses an arn to reference it? This would be most flexible and the arn can be specified either as fully qualified string, or a reference (Ref, Getatt or some Join construct)

asprouse commented:
@eahefnawy the issue with your suggestion is that `S3Bucket{normalizedBucketName}` is a moving target if you are generating your bucket name  dynamically. For example this is the naive way to do what you're suggesting and was suggested in #2717:
```yml
functions:
 upload-hook:
    handler: api/upload-hook/index.handler
    events:
      - s3:
          bucket: ${self:service}.${self:custom.stage}.assets
          event: s3:ObjectCreated:*

resources:
  Resources:
    S3Bucket${normalizedBucketName}:
      Type: AWS::S3::Bucket
      Properties:
        AccessControl: Private
        BucketName: "${self:service}.${self:custom.stage}.assets"
        CorsConfiguration:
          CorsRules:
          - AllowedOrigins:
              - http://example.dev:3001
            AllowedHeaders:
              - "*"
            AllowedMethods:
              - PUT
              - GET
          - AllowedOrigins:
              - https://dev.example.com
            AllowedHeaders:
              - "*"
            AllowedMethods:
              - PUT
              - GET
```
Is impossible with the current setup as you mentioned because variables are not parsed in config keys. The only solution that I see for this use case is to move all of the custom config variables outside of the the `serverless.yml` and implement a totally custom `serverless-resources.js` in which you define your `resources` section so that you can generate your `S3Bucket${normalizedBucketName}` object key dynamically. An issue with that solution is that there is currently no way to access `self` variables in the `.js` configuration and to work around that those variables must be moved into a `.js` config file. 

In the end it's a messy solution for a pretty common use case for S3. In our case it is to allow user uploads from the browser to S3! You are going to run into this issue across cloud providers for example you also need to configure CORS using [Google Cloud Storage Buckets](https://cloud.google.com/storage/docs/cross-origin).

As far as it being technical debt I am very sympathetic, I am all for developing a generic solution. The problem is that Serverless insists on auto-creating resources for the user. There should be a generic **straightforward** mechanism to configure these resources or, as @nachoab suggests, simply disabling the auto-creation.

johnrjj commented:
Having similar issue -- Trying to create an S3 bucket w/ an event _and_ CORS enabled. There are work arounds, but I see this configuration type as more of a common case than an edge case. Would love to see this solved w/out too much ceremony if possible.
